walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,122 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    //TODO (@zous): Clean up comments.
+
+  // Runtime: O(n^2) where n is number of events passed in. 
+  // Ideally this problem could be dealt in O(n) time with the events sorted before 
+  // finding timeslots. 
+  public Collection<TimeRange> query (Collection<Event> events, MeetingRequest request){
+        // For optional attendees, just run query twice. First with optional  attendees included, if that is 
+        // not empty then return. If empty, then query just with mandatory attendees. 
+      Collection<String> allAttendees = new ArrayList<>();
+      allAttendees.addAll(request.getAttendees());
+      allAttendees.addAll(request.getOptionalAttendees());
+      MeetingRequest requestAll = new MeetingRequest (allAttendees, request.getDuration()); 
+      Collection<TimeRange> solutionsAll = queryHelper(events, requestAll);
+      // Return query of with optional guests, if there are options
+      // or if there are no mandatory attendees.
+      if (!(solutionsAll.isEmpty()) || request.getAttendees().isEmpty()){
+        return solutionsAll;
+      }
+      else{
+          return queryHelper(events, request);
+      }
+  }
+
+  //Runtime: O(n^2) where n is the size of the events. 
+  private Collection<TimeRange> queryHelper(Collection<Event> events, MeetingRequest request) {
```

Comment:
What exactly is |queryHelper| doing here? Just removing invalid time ranges? Maybe it should be renamed to |removeInvalidTimeRanges| or |removeInvalidTimes|?

Reply:
queryhelper was actually the original query function before we were asked to consider optional attendees.